### PR TITLE
Ensure that delete jobs run when projects are deleted (#1893)

### PR DESCRIPTION
**Pull request by thedadams**
There are a few situations where we need a buffer between a project being deleted and the corresponding namespace being deleted. One such scenario is ensuring that delete jobs run when a project is deleted.

This change adds a project instance, which is this buffer. Using the project instance, we can ensure that all apps are successfully cleaned up before the namespace is deleted. This should also give the user the ability to use the ignore-cleanup functionality when deleting a project.

### Checklist
- [x] The title of this PR would make a good line in Acorn's Release Note's Changelog
- [x] The title of this PR ends with a link to the main issue being address in parentheses, like: `This is a title (#1216)`. [Here's an example](https://github.com/acorn-io/runtime/pull/1199)
- [x] All relevant issues are referenced in the PR description. *NOTE: don't use [GitHub keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) that auto-close issues*
- [x] Commits follow [contributing guidance](https://github.com/acorn-io/runtime/blob/main/CONTRIBUTING.md#commits)
- [x] Automated tests added to cover the changes. If tests couldn't be added, an explanation is provided in the Verification and Testing section
- [x] Changes to user-facing functionality, API, CLI, and upgrade impacts are clearly called out in PR description
- [ ] PR has at least two approvals before merging (or a reasonable exception, like it's just a docs change)

Requires: https://github.com/acorn-io/baaah/pull/90
Issue: https://github.com/acorn-io/runtime/issues/1893

### CODE REVIEW

File: pkg/controller/defaults/testdata/memory/all-set-overwrite/expected.golden [hunk 1]
```diff
@@ -40,16 +40,13 @@ status:
   appStatus: {}
   columns: {}
   conditions:
-    reason: Success
-    status: "True"
-    success: true
+  - error: true
```
Comment (thedadams) [inline pkg/controller/defaults/testdata/memory/all-set-overwrite/expected.golden:43]: I need to fix this.

File: pkg/controller/appdefinition/testdata/computeclass/all-set/input.yaml [hunk 2]
```diff
@@ -10,9 +10,6 @@ spec:
     "": sample-compute-class
 status:
   observedGeneration: 1
-  defaults:
```
Comment (thedadams) [inline pkg/controller/appdefinition/testdata/computeclass/all-set/input.yaml:13]: This was a duplicate entry.

File: pkg/controller/defaults/testdata/region/default/expected.golden [hunk 3]
```diff
@@ -21,15 +21,13 @@ status:
...
+      *v1.ProjectInstance "app-namespace" not found'
+    reason: Error
+    status: "False"
     type: defaults
-  defaults:
-    memory:
-      "": 0
-      container-name: 0
-    region: local
+  defaults: {}
```
Comment (g-linville) [inline pkg/controller/defaults/testdata/region/default/expected.golden:30]: It looks like this and several other tests went from expecting success to expecting failure. Was this intended?

File: pkg/server/registry/apigroups/acorn/projects/storage.go [hunk 4]
```diff
@@ -1,37 +1,76 @@
...
 		WithValidateCreate(validator).
 		WithValidateUpdate(validator).
 		WithTableConverter(tables.ProjectConverter).
 		Build()
 }
+
+type projectCreater struct {
+	creater strategy.Creater
+	client  kclient.Client
+}
```
Comment (g-linville) [inline pkg/server/registry/apigroups/acorn/projects/storage.go:41]: nit: I know the underlying typo/misspelling comes from mink, but I think this is worth changing anyway

```suggestion
type projectCreator struct {
	creator strategy.Creater
	client  kclient.Client
}
```

(Would need to modify other parts of this file too.)

File: pkg/uninstall/uninstall.go [hunk 5]
```diff
@@ -228,6 +238,28 @@ func Uninstall(ctx context.Context, opts *Options) error {
 		}
 	}
 
+	// Delete projects first. This will allow the acorn-controller to do all the cleanup it needs to do.
+	pterm.Info.Println("Deleting projects")
+	projectList := new(apiv1.ProjectList)
+	projects := make([]kclient.Object, 0, len(projectList.Items))
```
Comment (g-linville) [inline pkg/uninstall/uninstall.go:244]: nit: won't `len(projectList.Items)` always be 0 here?

File: pkg/controller/defaults/testdata/region/default/expected.golden (hunk 3 above)
Comment (thedadams) [reply pkg/controller/defaults/testdata/region/default/expected.golden:30]: I thought I fixed all of these, my apologies. I will update.

File: integration/helper/project.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TempProject(t *testing.T, client client.WithWatch) *v1.ProjectInstance {
+	t.Helper()
+	return tempCreateHelper(t, client)
+}
+
+func tempCreateHelper(t *testing.T, client client.WithWatch) *v1.ProjectInstance {
```
[the remark this answers is not in the document]
Comment (thedadams) [reply integration/helper/project.go:21]: There was another function that was using the helper, but I removed it (it wasn't necessary). I will simplify this.

File: pkg/uninstall/uninstall.go (hunk 5 above)
Comment (thedadams) [reply pkg/uninstall/uninstall.go:244]: Yes, will update.

**Review by g-linville — APPROVED** [review]

File: pkg/uninstall/uninstall.go [hunk 7]
```diff
@@ -80,6 +82,14 @@ func baseResources(ctx context.Context, c kclient.Client) (resources []kclient.O
 		}
 	}
 
+	controllerLease := &coordinationv1.Lease{}
+	if err := c.Get(ctx, kclient.ObjectKey{Namespace: "kube-system", Name: "acorn-controller"}, controllerLease); !apierror.IsNotFound(err) && err != nil {
+		return nil, err
+	}
+	if err == nil {
```
Comment (StrongMonkey) [inline pkg/uninstall/uninstall.go:89]: the err here is not the err in the if statement above?

File: pkg/uninstall/uninstall.go [hunk 8]
```diff
@@ -80,6 +82,14 @@ func baseResources(ctx context.Context, c kclient.Client) (resources []kclient.O
 		}
 	}
 
+	controllerLease := &coordinationv1.Lease{}
+	if err := c.Get(ctx, kclient.ObjectKey{Namespace: "kube-system", Name: "acorn-controller"}, controllerLease); !apierror.IsNotFound(err) && err != nil {
```
Comment (StrongMonkey) [inline pkg/uninstall/uninstall.go:86]: Does the lease live in acorn-system or kube-system?

File: pkg/uninstall/uninstall.go (hunk 7 above)
Comment (thedadams) [reply pkg/uninstall/uninstall.go:89]: You're right. Will fix.

File: pkg/uninstall/uninstall.go (hunk 8 above)
Comment (thedadams) [reply pkg/uninstall/uninstall.go:86]: It lives in the `kube-system` namespace.

**Review by StrongMonkey — APPROVED** [review]

**Pull request merged**